File: app/schemas/friend_activity.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.models.schedule import DaySessionType


class ActivityFeedEntryRead(BaseModel):
    """One outbox_events row (level_up, training_completed, or
    party_completed), reshaped for display -- see
    FriendActivityService.get_feed. type-specific fields (level,
    session_type, party_size) are None when they don't apply to this entry's
    event_type, rather than splitting into three response schemas.
    """

    id: uuid.UUID
    event_type: Literal["level_up", "training_completed", "party_completed"]
    user_id: uuid.UUID
    first_name: str
    last_name: str
    avatar_url: str | None = None
    created_at: datetime
    level: int | None = None
    session_type: DaySessionType | None = None
    party_size: int | None = None

File: frontend/src/types/friendActivity.ts
import type { DaySessionType } from './schedule'

export type ActivityFeedEventType = 'level_up' | 'training_completed' | 'party_completed'

// GET /friends/feed -- one row per outbox_events row (level_up,
// training_completed, or party_completed), reshaped for display.
// level/session_type/party_size are only set for the matching event_type.
export interface ActivityFeedEntryRead {
  id: string
  event_type: ActivityFeedEventType
  user_id: string
  first_name: string
  last_name: string
  avatar_url: string | null
  created_at: string
  level: number | null
  session_type: DaySessionType | null
  party_size: number | null
}
